src/dom-utils.ts

```diff
@@ -27,55 +27,77 @@ export function moveFocusToDialog(el: HTMLElement) {
   focused.focus()
 }
 
-/**
- * Get focusable children by recursively traversing the subtree of the given
- * element, while accounting for Shadow DOM subtrees
- */
-export function getFocusableChildren(el: ParentNode): HTMLElement[] {
-  // Check for the base case of our recursion
-  if (el instanceof HTMLElement) {
-    // If this element is marked as inert, neither it nor any member of its
-    // subtree will be focusable, therefore there are no focusable children
-    if (el.inert) return []
+// Get the first and last focusable elements in a given subtree
+export function getFirstAndLastFocusableChild(el: HTMLElement) {
+  const first = findFocusableElement(el, true)
+  let last = null
+  if (first !== null) {
+    last = findFocusableElement(el, false) || first
+  }
+  return [first, last]
+}
 
-    // If this element has no children or if it nullifies its children’s
-    // semantics, there are no focusable children
-    if (!el.firstElementChild || el.matches(PRESENTATIONAL_CHILDREN_SELECTOR)) {
-      // If the element itself is focusable, return it otherwise return nothing
-      return isFocusable(el) ? [el] : []
-    }
+function findFocusableElement(
+  node: HTMLElement,
+  forward: boolean
+): HTMLElement | null {
+  // If we're walking forward, check if this node is focusable,
+  // and return it immediately if it is.
+  if (forward && isFocusable(node)) {
+    return node
   }
 
-  let focusableEls: HTMLElement[] = []
+  // If this node can't have focusable children, we can't go any deeper
+  // regardless of whether we're walking forward or backward.
+  if (!canHaveFocusableChildren(node)) {
+    return null
+  }
 
-  // Walk all the immediate children of this element (with some type casting
-  // because `el.children` is an `HTMLCollection`)
-  for (const curr of el.children as unknown as HTMLElement[]) {
-    // If this element has a Shadow DOM attached, check the Shadow subtree for
-    // focusable children
-    if (curr.shadowRoot) {
-      focusableEls = focusableEls.concat(getFocusableChildren(curr.shadowRoot))
+  // Start walking the DOM tree, looking for focusable elements.
+  // Case 1: If this node has a shadow root, search it recursively.
+  if (node.shadowRoot) {
+    const shadowRoot = node.shadowRoot
+    // Descend into this subtree.
+    let child = forward
+      ? shadowRoot.firstElementChild
+      : shadowRoot.lastElementChild
+    // Traverse siblings, searching the subtree of each one
+    // for focusable elements.
+    while (child) {
+      const focusableEl = findFocusableElement(child as HTMLElement, forward)
+      if (focusableEl) return focusableEl
+      child = forward ? child.nextElementSibling : child.previousElementSibling
```

Comment:
Since we repeat this a bit, maybe these utils could make reading the core logic a bit easier?

```js
const getStartingEl = (el: HTMLElement, forward: boolean): Element | null =>
  forward ? el.firstElementChild : el.lastElementChild

const getNextEl = (el: HTMLElement, forward: boolean): Element | null =>
  forward ? el.nextElementSibling : el.previousElementSibling
```

Reply:
I like that idea!